containers/src/Data/Map/Internal.hs

```diff
@@ -4327,11 +4333,16 @@ sameSizeLiftEq2 keq eq m1 m2 =
       Nothing -> False :*: it
       Just (KeyValue ky y :*: it') -> (keq kx ky && eq x y) :*: it'
 {-# INLINE sameSizeLiftEq2 #-}
+#else
+instance (Eq k,Eq a) => Eq (Map k a) where
+  t1 == t2  = (size t1 == size t2) && (toAscList t1 == toAscList t2)
+#endif
```

Comment:
I'd prefer if we did not have entirely separate definitions (without good reason).  
The trouble here is no Eq1/Eq2, so we can instead lift the `liftEq2` definition to the top level and use it for MHS.  
Same for Ord1/Ord2 and other similar instances.